### ISSUE TITLE
SetLocale is not working as expected - Instance/Babel problem 

### ISSUE THREAD

**Issue by teteupoleza** [jquense/yup#293]
## Problem
I'm trying to use the function `setLocale` and It's not translating the errors. The problem seems to be with the library instance, some babel configuration or `import` problem.  

When I use setLocale and return the same instance of `yup` to my validation files It does work as expected. 

## Scenario 
**Should work like this. Right?**
index.js
```js
import { setLocale } from 'yup';

setLocale({
  mixed: {
    required: 'Preencha esse campo para continuar'
  },
  string: {
    email: 'Preencha um e-mail válido',
    min: 'Valor muito curto (mínimo ${min} caracteres)',
    max: 'Valor muito longo (máximo ${max} caracteres)'
  },
  number: {
    min: 'Valor inválido (deve ser maior ou igual a ${min})',
    max: 'Valor inválido (deve ser menor ou igual a ${max})'
  }
});
```

personal-data-validations.js
```js
import * as yup from 'yup';

const PersonalDataSchema = yup.object().shape({
  fullName: yup.string().min(5).max(70).required()
});

export { PersonalDataSchema };
```

## The workaround 
I created a index to my validations -> `validations.js` and export yup from there. After that I imported the yup instance from my validations index and now **It's working**.

validations.js
```js
import * as yup from 'yup';

yup.setLocale({
  mixed: {
    required: 'Preencha esse campo para continuar'
  },
  string: {
    email: 'Preencha um e-mail válido',
    min: 'Valor muito curto (mínimo ${min} caracteres)',
    max: 'Valor muito longo (máximo ${max} caracteres)'
  },
  number: {
    min: 'Valor inválido (deve ser maior ou igual a ${min})',
    max: 'Valor inválido (deve ser menor ou igual a ${max})'
  }
});

export default yup;
```

personal-data-validations.js
```js
import yup from 'validation';

const PersonalDataSchema = yup.object().shape({
  fullName: yup.string().min(5).max(70).required()
});

export { PersonalDataSchema };
```

**Comment by teteupoleza** [comment]
Maybe It's related to -> https://github.com/jquense/yup/issues/260

**Comment by jquense** [comment]
you need to set the locale before you import yup elsewhere. schema constructed before you use setLocale will have the old values

**Comment by teteupoleza** [comment]
@jquense I did not use setLocale before. I put it in my `index.js`as mentioned and this file is load before any validations file.

**Comment by Rockson** [comment]
For me setLocale is not working at all. Here is the sandbox: https://codesandbox.io/s/jj8zzp2mn5
Any clue on this?

**Comment by medeirosmarcus** [comment]
> For me setLocale is not working at all. Here is the sandbox: https://codesandbox.io/s/jj8zzp2mn5
> Any clue on this?

@Rockson You are setting the locale to number but trying to validade a string. Correct the types and it works.

**Comment by naglik** [comment]
@teteupoleza did you solve your problem? Im gonna use your workaround, but I would really like to setLocale in a central place of my next.js app (and somehow combine it with next-i18next)..if such thing is even possible

**Comment by std4453** [comment]
@jquense Well I wonder whether is there a way to use `yup` *instances* instead of importing it directly.

**Comment by MarcoScabbiolo** [comment]
If anyone has this problem using TypeScript you have to define all your schemas as a function that takes yup, for example:

```typescript
import * as yup from 'yup';

export default (y: typeof yup) =>
  y
    .string()
    .email()
    .max(300);
```

Pass yup (`import * as yup from 'yup'`) **after** calling `setLocale` on it.

**Comment by hyl0327** [comment]
I had the same problem.

After some research, I found the cause was because the module actually using yup (in your case, personal-data-validations.js) was being run before the line of code calling setLocale (in your case, that line is in index.js), so yup saw the original locale.

**Comment by joris-** [comment]
Hi ! 
Having the same problem, I use this work around _(probably not better)_ :

```
// webpack.config.js
const { NormalModuleReplacementPlugin } = require('webpack');

module.exports = {
    ...,
    plugins: [
        new NormalModuleReplacementPlugin(
            /yup\/lib\/locale\.js/,
            path.resolve(process.cwd(), './src/utils/custom-yup-locale.js'),
        ),
   ]
}
```

It will replace `locale.js` of yup by your custom

**Comment by samuelpetroline** [comment]
I know this issue is closed but the error still happens. I've handled it by delaying the main code execution:

```js
import { registerRootComponent } from 'expo'
import { initLocale } from './src/Schemas' // Method where 'setLocale' is called

const loadApp = async () => {
  const App = await import('./src/App')

  registerRootComponent(App.default)
}

initLocale()
loadApp()
```

There may be better ways to handle this but it worked for me and may also help someone else.

**Comment by RadovanPelka** [comment]
> If anyone has this problem using TypeScript you have to define all your schemas as a function that takes yup, for example:
> 
> ```ts
> import * as yup from 'yup';
> 
> export default (y: typeof yup) =>
>   y
>     .string()
>     .email()
>     .max(300);
> ```
> 
> Pass yup (`import * as yup from 'yup'`) to your schemas **after** calling `setLocale` on it.

please full example

**Comment by soulchild** [comment]
> you need to set the locale before you import yup elsewhere. schema constructed before you use setLocale will have the old values

What if I want to share a yup schema defined in a module between backend and frontend and have a different locale for the frontend? I'd need to inject a yup instance into all schemas. It'd be nice if there was a way to alter the locale *after* the schema has been declared - maybe as an option to `validate`?

**Comment by kfrankot** [comment]
In case anyone runs into this, a reminder that you can pass functions to the setLocale definitions.  For example, if you needed to dynamically change language, you can do something like this:

```
import * as yup from 'yup'
import { getCurrentLanguage } from '../utils' 

yup.setLocale({
  mixed: {
    required: () => {
      const lang = getCurrentLanguage()
      if(lang === 'en-US') {
        return 'Field required'
      }
      // etc...
    }
  }
})
```

**Comment by sgarcia-dev** [comment]
There's an open PR since May 22 that addresses this issue, @jquense could you take a look?
https://github.com/jquense/yup/pull/1371

**Comment by josejaviersanahuja** [comment]
I would like to show my workaround as I considere I didn't understand your workarounds.

Basically, I was working on a LanguageContext. 

`const LanguageContext = createContext({
  lang: ESPAÑOL,
  setLanguage: () => { },
  Yup: null,
});

export function useLanguage() {
  return useContext(LanguageContext);
}

export function LanguageProvider({ children }) {
  const [lang, setLanguage] = useState(ESPAÑOL);
  const [Yup, setYup] = useState(null)
  const valueContext = {
    lang,
    setLanguage,
    Yup,
  };

  useEffect(() => {
    // yup
    setLocale(errorsYup[lang]) // errorsYup is my custom error mapping
    import('yup').then(yup => {
      setYup(yup)
    })
    
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [lang])
`
So basically, I import Yup from my context from now on. It doesn't change automatically, but as soon as the component rerenders, it gets the new values.

If you think my workaround is bad permormant or a bad solution or you have a better way, please share how to improve it

**Comment by sebastian-budka-mir** [comment]
I also encountered the same problem in our code, in main tsx file we imported yup initializer file after other components. Solution was to make sure that yup setLocale initializer file was imported before all other components (ex. AppRouter)

**Comment by thany** [comment]
Sorry, why has this been fixed yet? It's 2024 and the bug is still there, 6 years after describing it and making a workaround.

Please reopen @jquense
